Layer0 - v4

jeffpatzer opened the pull request:
*Issue #, if available:*
None

*Description of changes:*
- Upgrades Layer0 version from original pull request (https://github.com/aws-samples/retail-demo-store/pull/223)
- Pulls and merges in updates from time original pull request

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

ronaks4 commented:
testing based on https://github.com/moovweb-demos/retail-demo-store/blob/layer0/src/web-ui#site-optimization-with-layer0

BastLeblanc commented on README.md:
```diff
@@ -14,6 +14,10 @@ The architecture is supported by several managed services including [Amazon Cogn
 
 ![Retail Demo Store Architecture](./workshop/images/retaildemostore-architecture.png)
 
+## Site Optimization with Layer0
```
I would prefer to see this in the Partner integration chapter to keep it consistent

BastLeblanc commented on aws/cloudformation-templates/base/_template.yaml:
```diff
@@ -2,12 +2,12 @@
   CreateElasticsearchServiceLinkedRole: !Equals
     - !Ref CreateElasticsearchServiceLinkedRole
-    - 'Yes'
+    - "Yes"
```
I suppose all these single to double quotes changes come from cfn-lint. which is good.
Was it ran on every cfn template or just the one impacted by the PR?

It makes a big PR, hard to decipher from the real changes.

jeffpatzer replied:
Yes, sorry it came from the Linter in my editor. Only run on those impacted by the PR during a save event.

BastLeblanc commented on src/web-ui/build.sh:
```diff
@@ -0,0 +1,11 @@
+# It generates ./dist output folder of web-ui Vue application.
+# If $LAYER0_DEPLOY_TOKEN is present it generates ./.layer0 output folder for deployment on Layer0.
+
+if [ $LAYER0_DEPLOY_TOKEN = "NONE" ]; then
```
is it NONE or empty?

because in src/web-ui/.env.template there is:
LAYER0_DEPLOY_TOKEN=

which seems to be a default to empty?

jeffpatzer replied:
Should be `NONE` by default. I added that.

BastLeblanc commented on workshop/7-PerformanceOptimization/7.1-Layer0.ipynb:
```diff
@@ -0,0 +1,139 @@
+{
```
it should be numbered 8 for the workshop
(7 is location services)

jeffpatzer replied:
Changed.

BastLeblanc reviewed (commented):
Please see comment inline.

General comments/remarks
- it misses a high level demo guide, getting started (cloudformation template parameters, what to demo from a "user" perspective)
- explain in a few words the value proposition from Layer0

jeffpatzer commented:
@BastLeblanc thank you for the PR review 🙌 

I addressed your comments inline and also the general ones here. 


> Please see comment inline.
> 
> General comments/remarks
> 
> * it misses a high level demo guide, getting started (cloudformation template parameters, what to demo from a "user" perspective)

Moved details, including a Demo, into the notebook instructions and pulled them from the readme of the `web-ui` folder. 

> * explain in a few words the value proposition from Layer0

I think this is in the first line of the notebook? There is a paragraph detailing why you would use Layer0. Should I change or expand on it?

ronaks4 commented:
I walk through the instructions with Jeff. @jeffpatzer will add cert upload for Loadbalancer instructions with some more details. Thanks for your review @BastLeblanc

jeffpatzer commented:
@ronaks4 instructions for SSL cert generation have been added

ronaks4 commented:
deployed it at https://ronaks4-retail-demo-aws-example-com-default.layer0-limelight.link/#/ and tested.

BastLeblanc reviewed (changes requested):
Hi,

I tried to test the whole PR, but went into issues:

Still some issues before we can merge this PR:
1) where is the doc linked here?
Check out [web-ui](./src/web-ui#site-optimization-with-layer0) for the guide

2) The PR has conflict, it's up to you to fix them before the PR can be merged

3) The instructions are confusing(or missing?), in the workshop it says:
"Make sure you have Retail Demo Store (RDS) deployed in your AWS account before proceeding. "

But when we deploy the cfn it asks for a tokenid from Layer0.
You have to make it clear on how NOT to deploy the layer0 specifics. leaving the tokenid to empty is enough? -> to be written in the CFN

BastLeblanc commented:
also, you link to - [AWS Retail Store on Layer0 demo](https://layer0-docs-layer0-aws-store-example-default.layer0-limelight.link)

Who's maintaining this demo?

BastLeblanc commented:
Bug in gen_env.sh :

```
$ ./gen_env.sh
./gen_env.sh: line 57: [: =: unary operator expected
```

if the TOKEN is empty, need to be protected:
`if [ "$LAYER0_DEPLOY_TOKEN" = "NONE" ]; then`

ronaks4 commented:
I will update and let you know. thank you.

BastLeblanc commented:
work in progress which may end be from a different PR by @ronaks4 for simplicity (and we would close this one)

jeffpatzer commented:
> Hi,
> 
> I tried to test the whole PR, but went into issues:
> 
> Still some issues before we can merge this PR:
> 
> 1. where is the doc linked here?
>    Check out [web-ui](./src/web-ui#site-optimization-with-layer0) for the guide
>    You should prepare 2 location for the instructions, one in a readme, one in the workshop
> 2. The PR has conflicts, it's up to you to fix them before the PR can be merged
> 3. The instructions are confusing(or missing?), in the workshop it says:
>    "Make sure you have Retail Demo Store (RDS) deployed in your AWS account before proceeding. "
> 
> But when we deploy the cfn it asks for a tokenid from Layer0. You have to make it clear on how NOT to deploy the layer0 specifics. leaving the tokenid to empty is enough? -> to be written in the CFN
> 
> 4. if https is a requirement, it should be fixed upstream, not in a workshop. Customers who want to test layer0 shouldn't be bothering about changing the core of the Retail Demo Store to https

1. CHanged link location. 
2. Fixed the conflicts
3. Deploy token removed from PR as a whole. Instructions now reflect separate usage. 
4. Agreed. I'm not sure how to configure the Cloud Deploy script to do that though. I think @ronaks4 was going to issue another PR with those updates?


> Bug in gen_env.sh :
> 
> ```
> $ ./gen_env.sh
> ./gen_env.sh: line 57: [: =: unary operator expected
> ```
> 
> if the TOKEN is empty, need to be protected: `if [ "$LAYER0_DEPLOY_TOKEN" = "NONE" ]; then`

Fixed. 



> Some further remarks on the workshop:
> 
> 1. you link to - [AWS Retail Store on Layer0 demo](https://layer0-docs-layer0-aws-store-example-default.layer0-limelight.link)
> 
> Who's maintaining this demo?
> 
> 2. first instruction is
> 
> 1. Make sure your terminal is open in `%project_root%/src/web-ui` folder
> 
> > which terminal? local? cloud9?
> > I would guess it's local on our machine, then you have to explain it (or better advise for a cloud9 env).
> 
> 3. I think the whole concept would benefit from an architecture diagram, it's hard to understand which component is used locally, on AWS, on layer0

- It's maintained on the Layer0 side. 
- Specified the terminal

jeffpatzer commented:
@BastLeblanc changes in per your comments
@ronaks4 changes in per your comments

BastLeblanc reviewed (approved):
the PR is approved. 
As discussed the AWS_RDS.jpg architecture could be better and will be updated in a subsequent PR asap.
